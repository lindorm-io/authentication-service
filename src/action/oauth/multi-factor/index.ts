export * from "./multi-factor-challenge";
export * from "./multi-factor-token";
