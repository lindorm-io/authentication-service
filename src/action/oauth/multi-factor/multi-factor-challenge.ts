import Joi from "@hapi/joi";
import { APIError, HttpStatus } from "@lindorm-io/core";
import { IAuthContext } from "../../../typing";
import { InvalidClientError } from "@lindorm-io/koa-client";
import { InvalidPermissionError, InvalidSubjectError } from "../../../error";
import { JOI_CHALLENGE_TYPE, JOI_EMAIL, JOI_GRANT_TYPE } from "../../../constant";
import { MultiFactorChallengeType } from "../../../enum";
import { assertSessionIsNotExpired } from "../../../support";
import { isLocked } from "@lindorm-io/jwt";

export interface IPerformMultiFactorChallengeOptions {
  authenticatorId?: string;
  challengeType: string;
  grantType: string;
  subject: string;
}

export interface IPerformMultiFactorChallengeData {
  bindingMethod?: string;
  challengeType: string;
  oobCode?: string;
}

const schema = Joi.object({
  authenticatorId: Joi.string(),
  challengeType: JOI_CHALLENGE_TYPE,
  grantType: JOI_GRANT_TYPE,
  subject: JOI_EMAIL,
});

export const performMultiFactorChallenge = (ctx: IAuthContext) => async (
  options: IPerformMultiFactorChallengeOptions,
): Promise<IPerformMultiFactorChallengeData> => {
  await schema.validateAsync(options);

  const { client, repository, token } = ctx;
  const { challengeType, subject } = options;
  const {
    multiFactor: { subject: sessionId },
  } = token;

  const session = await repository.session.find({ id: sessionId });

  assertSessionIsNotExpired(session);

  if (session.clientId !== client.id) {
    throw new InvalidClientError(client.id);
  }

  if (session.authorization.email !== subject) {
    throw new InvalidSubjectError(subject);
  }

  const account = await repository.account.find({ email: subject });

  if (isLocked(account.permission)) {
    throw new InvalidPermissionError();
  }

  switch (challengeType) {
    case MultiFactorChallengeType.OTP:
      return {
        challengeType,
        bindingMethod: "prompt",
      };

    case MultiFactorChallengeType.OOB:
      throw new APIError("Challenge type is not supported", {
        statusCode: HttpStatus.ClientError.FORBIDDEN,
      });

    default:
      throw new APIError("Unknown challenge type", {
        statusCode: HttpStatus.ClientError.FORBIDDEN,
      });
  }
};
