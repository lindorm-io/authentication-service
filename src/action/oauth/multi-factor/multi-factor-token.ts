import Joi from "@hapi/joi";
import { IAuthContext, ICreateTokensData } from "../../../typing";
import { InvalidClientError } from "@lindorm-io/koa-client";
import { InvalidPermissionError, InvalidSubjectError } from "../../../error";
import { JOI_EMAIL, JOI_GRANT_TYPE } from "../../../constant";
import { assertAccountOTP, assertSessionIsNotExpired, authenticateSession, createTokens } from "../../../support";
import { isLocked } from "@lindorm-io/jwt";

export interface IPerformMultiFactorTokenOptions {
  bindingCode: string;
  grantType: string;
  subject: string;
}

const schema = Joi.object({
  bindingCode: Joi.string().required(),
  grantType: JOI_GRANT_TYPE,
  subject: JOI_EMAIL,
});

export const performMultiFactorToken = (ctx: IAuthContext) => async (
  options: IPerformMultiFactorTokenOptions,
): Promise<ICreateTokensData> => {
  await schema.validateAsync(options);

  const { client, repository, token } = ctx;
  const { bindingCode, subject } = options;
  const {
    multiFactor: { authMethodsReference, subject: sessionId },
  } = token;

  const session = await repository.session.find({ id: sessionId });

  assertSessionIsNotExpired(session);

  if (session.clientId !== client.id) {
    throw new InvalidClientError(client.id);
  }

  if (session.authorization.email !== subject) {
    throw new InvalidSubjectError(subject);
  }

  const account = await repository.account.find({ email: subject });

  if (isLocked(account.permission)) {
    throw new InvalidPermissionError();
  }

  await assertAccountOTP(account, bindingCode);

  const authenticated = await authenticateSession(ctx)({
    account,
    session,
  });

  return createTokens(ctx)({
    account,
    authMethodsReference: `${authMethodsReference} otp`,
    client,
    responseType: session.authorization.responseType,
    session: authenticated,
  });
};
