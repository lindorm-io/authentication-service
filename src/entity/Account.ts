import { AccountEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { Permission } from "@lindorm-io/jwt";

export interface IAccountOTP {
  signature: string;
  uri: string;
}

export interface IAccountPassword {
  signature: string;
  updated: Date;
}

export interface IAccount extends IEntity {
  email: string;
  identityLinked: boolean;
  otp: IAccountOTP;
  password: IAccountPassword;
  permission: string;
}

export interface IAccountOptions extends IEntityBaseOptions {
  email: string;
  identityLinked?: boolean;
  otp?: IAccountOTP;
  password?: IAccountPassword;
  permission?: string;
}

export class Account extends EntityBase implements IAccount {
  private _email: string;
  private _identityLinked: boolean;
  private _otp: IAccountOTP;
  private _password: IAccountPassword;
  private _permission: string;

  constructor(options: IAccountOptions) {
    super(options);
    this._email = options.email.toLowerCase();
    this._identityLinked = options.identityLinked || false;
    this._otp = {
      signature: options.otp?.signature || null,
      uri: options.otp?.uri || null,
    };
    this._password = {
      signature: options.password?.signature || null,
      updated: options.password?.updated || null,
    };
    this._permission = options.permission || Permission.USER;
  }

  public get email(): string {
    return this._email;
  }
  public set email(email: string) {
    this._email = email.toLowerCase();
    this.addEvent(AccountEvent.EMAIL_CHANGED, { email: this._email });
  }

  public get identityLinked(): boolean {
    return this._identityLinked;
  }
  public set identityLinked(identityLinked: boolean) {
    this._identityLinked = identityLinked;
    this.addEvent(AccountEvent.EMAIL_CHANGED, { identityLinked: this._identityLinked });
  }

  public get otp(): IAccountOTP {
    return this._otp;
  }
  public set otp(otp: IAccountOTP) {
    this._otp = otp;
    this.addEvent(AccountEvent.OTP_CHANGED, { otp: this._otp });
  }

  public get password(): IAccountPassword {
    return this._password;
  }
  public set password(password: IAccountPassword) {
    this._password = password;
    this.addEvent(AccountEvent.PASSWORD_CHANGED, { password: this._password });
  }

  public get permission(): string {
    return this._permission;
  }
  public set permission(permission: string) {
    this._permission = permission;
    this.addEvent(AccountEvent.PERMISSION_CHANGED, { permission: this._permission });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== AccountEvent.CREATED) continue;
      throw new EntityCreationError("Account");
    }
    this.addEvent(AccountEvent.CREATED, {
      email: this._email,
      permission: this._permission,
      password: this._password,
      otp: this._otp,
      created: this._created,
      updated: this._updated,
    });
  }
}
