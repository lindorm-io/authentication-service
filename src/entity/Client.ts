import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { ClientEvent } from "../enum";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  emailAuthorizationUri: string;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  emailAuthorizationUri?: string;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _emailAuthorizationUri: string;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._emailAuthorizationUri = options?.emailAuthorizationUri || null;
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get emailAuthorizationUri(): string {
    return this._emailAuthorizationUri;
  }
  public set emailAuthorizationUri(uri: string) {
    this._emailAuthorizationUri = uri;
    this.addEvent(ClientEvent.EMAIL_AUTHORIZATION_URI_CHANGED, { uri: this._emailAuthorizationUri });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new Error("Client has already been created");
    }
    this.addEvent(ClientEvent.CREATED, {
      name: this._name,
      emailAuthorizationUri: this._emailAuthorizationUri,
      approved: this._approved,
      description: this._description,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
