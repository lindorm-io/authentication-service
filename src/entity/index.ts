export * from "./Account";
export * from "./RequestLimit";
export * from "./Session";
