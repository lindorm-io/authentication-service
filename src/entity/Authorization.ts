import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { GrantType } from "../enum";
import { Scope } from "@lindorm-io/jwt";

export interface IAuthorization extends IEntity {
  challengeId: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export interface IAuthorizationOptions extends IEntityBaseOptions {
  challengeId?: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId?: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode?: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export class Authorization extends EntityBase implements IAuthorization {
  readonly challengeId: string;
  readonly clientId: string;
  readonly codeChallenge: string;
  readonly codeMethod: string;
  readonly deviceId: string;
  readonly email: string;
  readonly expires: Date;
  readonly grantType: GrantType;
  readonly otpCode: string;
  readonly redirectUri: string;
  readonly responseType: string;
  readonly scope: Array<Scope>;

  constructor(options: IAuthorizationOptions) {
    super(options);

    this.challengeId = options.challengeId || null;
    this.clientId = options.clientId;
    this.codeChallenge = options.codeChallenge;
    this.codeMethod = options.codeMethod;
    this.deviceId = options.deviceId || null;
    this.email = options.email;
    this.expires = options.expires;
    this.grantType = options.grantType;
    this.otpCode = options.otpCode || null;
    this.redirectUri = options.redirectUri;
    this.responseType = options.responseType;
    this.scope = options.scope;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
