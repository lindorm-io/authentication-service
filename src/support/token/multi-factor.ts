import { Audience } from "../../enum";
import { Authorization } from "../../entity";
import { Client } from "@lindorm-io/koa-client";
import { IKoaAuthContext } from "../../typing";
import { ITokenIssuerSignData } from "@lindorm-io/jwt";
import { JWT_MULTI_FACTOR_TOKEN_EXPIRY } from "../../config";

export interface IGetMultiFactorTokenOptions {
  authMethodsReference: Array<string>;
  authorization: Authorization;
  client: Client;
}

export const getMultiFactorToken = (ctx: IKoaAuthContext) => (
  options: IGetMultiFactorTokenOptions,
): ITokenIssuerSignData => {
  const { logger, issuer, metadata } = ctx;
  const { authMethodsReference, client, authorization } = options;

  logger.debug("creating multi factor token", {
    authorization: authorization.id,
    client: client.id,
    device: metadata.deviceId,
  });

  return issuer.auth.sign({
    audience: Audience.MULTI_FACTOR,
    authMethodsReference: authMethodsReference,
    clientId: client.id,
    deviceId: metadata.deviceId,
    expiry: client?.extra?.jwtMultiFactorTokenExpiry || JWT_MULTI_FACTOR_TOKEN_EXPIRY,
    subject: authorization.id,
  });
};
