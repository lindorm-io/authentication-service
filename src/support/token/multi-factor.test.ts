import MockDate from "mockdate";
import { getMultiFactorToken } from "./multi-factor";
import { Client, Device, Session } from "../../entity";
import {
  MOCK_CLIENT_OPTIONS,
  MOCK_DEVICE_OPTIONS,
  MOCK_SESSION_OPTIONS,
  MOCK_EC_TOKEN_ISSUER,
  MOCK_LOGGER,
} from "../../test/mocks";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("getMultiFactorToken", () => {
  let getMockContext: any;

  let client: Client;
  let device: Device;
  let session: Session;

  beforeEach(() => {
    getMockContext = () => ({
      logger: MOCK_LOGGER,
      issuer: { tokenIssuer: MOCK_EC_TOKEN_ISSUER },
    });

    client = new Client(MOCK_CLIENT_OPTIONS);
    device = new Device(MOCK_DEVICE_OPTIONS);
    session = new Session(MOCK_SESSION_OPTIONS);
  });

  test("should return a multi-factor token", () => {
    expect(
      getMultiFactorToken(getMockContext())({
        authMethodsReference: "authMethodsReference",
        client,
        device,
        session,
      }),
    ).toStrictEqual({
      expires: 1577863800,
      expiresIn: 1800,
      id: "be3a62d1-24a0-401c-96dd-3aff95356811",
      level: undefined,
      token: expect.any(String),
    });
  });

  test("should return a multi-factor token without device", () => {
    expect(
      getMultiFactorToken(getMockContext())({
        authMethodsReference: "authMethodsReference",
        client,
        session,
      }),
    ).toStrictEqual({
      expires: 1577863800,
      expiresIn: 1800,
      id: "be3a62d1-24a0-401c-96dd-3aff95356811",
      level: undefined,
      token: expect.any(String),
    });
  });
});
