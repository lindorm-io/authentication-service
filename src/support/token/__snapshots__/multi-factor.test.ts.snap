// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMultiFactorToken should return a multi-factor token 1`] = `
Object {
  "expires": 1577863800,
  "expiresIn": 1800,
  "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "level": undefined,
  "token": Object {
    "amr": "authMethodsReference",
    "aud": "multi-factor",
    "cid": "9ba5e0c5-f2aa-4706-a27d-66369b547b0b",
    "did": "deviceId",
    "exp": 1577863800,
    "iat": 1577862000,
    "iss": "https://test.authentication.lindorm.io",
    "jti": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "nbf": 1577862000,
    "sub": "be3a62d1-24a0-401c-96dd-3aff95356811",
  },
}
`;
